rfcs/0000-longjmp-pof-annotation.md

```diff
@@ -0,0 +1,85 @@
+- Feature Name: `annotation-for-safe-longjmp`
+- Start Date: 2019-06-11
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+- Project group: [FFI-unwind][project-group]
+
+ [project-group]: https://github.com/rust-lang/project-ffi-unwind
+
+ <!-- XXX for now, content is copied from prior drafts of the "C unwind" RFC. -->
+
+# Motivation
+[motivation]: #motivation
+
+Additionally, there are libraries such as `rlua` that rely on `longjmp` across
+Rust frames; on Windows, `longjmp` is implemented via [forced
+unwinding][forced-unwinding]. The current `rustc` implementation makes it safe
+to `longjmp` across Rust [POFs][POF-definition] (frames without `Drop` types),
+but this is not formally specified in an RFC or by the Reference.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC specifies that, regardless of the platform or the ABI string (`"C"` or
+`"C unwind"`), any platform features that may rely on forced unwinding is
+defined behavior when all unwound frames are POFs
+
+As an example:
+
+```rust
+#[cancelable]
+fn foo<D: Drop>(c: bool, d: D) {
+  if c {
+    drop(d);
+  }
+  longjmp_if_true(c);
+}
+
+/// Calls `longjmp` if `c` is true; otherwise returns normally.
+extern "C" fn longjmp_if_true(c: bool);

```

Comment:
I think I would expect this function to have a `#[cancelable]` attribute.

Reply:
That does tend to make the name "cancelable" feel less good, though.

Reply:
Yes, I think it does make sense to require `extern` functions to be annotated as well. Fortunately, I think we have plenty of time to bikeshed a name for the annotation!